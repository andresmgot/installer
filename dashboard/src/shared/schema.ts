import { IBasicFormParam } from "./types";

export function retrieveBasicFormParams(schema: any) {
  // TBD
  return {
    username: {
      path: "wordpressUsername",
      value: "user",
    } as IBasicFormParam,
  };
}

export function setValue(values: string, path: string, newValue: any) {
  // TBD
  return values;
}
