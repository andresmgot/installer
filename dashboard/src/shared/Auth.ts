import Axios, { AxiosResponse } from "axios";
const AuthTokenKey = "kubeapps_auth_token";

export class Auth {
  public static getAuthToken() {
    return localStorage.getItem(AuthTokenKey);
  }

  public static setAuthToken(token: string) {
    return localStorage.setItem(AuthTokenKey, token);
  }

  public static unsetAuthToken() {
    return localStorage.removeItem(AuthTokenKey);
  }

  public static wsProtocols() {
    const token = this.getAuthToken();
    if (!token) {
      return [];
    }
    return [
      "base64url.bearer.authorization.k8s.io." + btoa(token).replace(/=*$/g, ""),
      "binary.k8s.io",
    ];
  }

  public static fetchOptions(): RequestInit {
    const headers = new Headers();
    headers.append("Authorization", `Bearer ${this.getAuthToken()}`);
    return {
      headers,
    };
  }

  // Throws an error if accessing the Kubernetes API returns an Unauthorized error
  public static async validate(token?: string) {
    try {
      let options = {};
      if (token) {
        options = { headers: { Authorization: `Bearer ${token}` } };
      }
      await Axios.get("api/kube/", options);
    } catch (e) {
      const res = e.response as AxiosResponse;
      if (res.status === 401 || res.status === 403) {
        throw new Error("invalid token");
      }
    }
  }
}

export const axios = Axios.create();
