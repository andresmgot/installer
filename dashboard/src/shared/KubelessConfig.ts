import axios from "axios";

import { IKubelessConfigMap } from "./types";

export default class Function {
  public static async get() {
    const config = await axios.get<IKubelessConfigMap>(Function.SelfLink);
    return config;
  }

  public static async getRuntimes() {
    const config = await this.get();
    return JSON.parse(config.data.data["runtime-images"]);
  }

  private static Name: string = "kubeless-config";
  private static Namespace: string = "kubeless";
  private static APIBase: string = "/api/kube";
  private static APIEndpoint: string = `${Function.APIBase}/api/v1`;
  private static SelfLink: string = `${Function.APIEndpoint}/namespaces/${
    Function.Namespace
  }/configmaps/${Function.Name}`;
}
