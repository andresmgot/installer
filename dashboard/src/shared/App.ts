import { axios } from "./Auth";
import { IAppConfigMap } from "./types";

export class App {
  public static async waitForDeletion(name: string) {
    const timeout = 30000; // 30s
    return new Promise((resolve, reject) => {
      const interval = setInterval(async () => {
        const { data: { items: allConfigMaps } } = await axios.get<{
          items: IAppConfigMap[];
        }>(this.getConfigMapsLink([name]));
        if (allConfigMaps.length === 0) {
          clearInterval(interval);
          resolve();
        }
      }, 500);
      setTimeout(() => {
        clearInterval(interval);
        reject(`Timeout after ${timeout / 1000} seconds`);
      }, timeout);
    });
  }

  public static async exists(releaseName: string) {
    const { data: { items: allConfigMaps } } = await axios.get<{
      items: IAppConfigMap[];
    }>(this.getConfigMapsLink([releaseName]));
    if (allConfigMaps.length === 0) {
      return false;
    }
    return true;
  }

  // getConfigMapsLink returns the URL for listing Helm ConfigMaps for the given
  // set of release names.
  public static getConfigMapsLink(releaseNames?: string[]) {
    let query = "";
    if (releaseNames) {
      query = `,NAME in (${releaseNames.join(",")})`;
    }
    return `/api/kube/api/v1/namespaces/kubeapps/configmaps?labelSelector=OWNER=TILLER${query}`;
  }
}
