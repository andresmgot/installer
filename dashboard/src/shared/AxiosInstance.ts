import { AxiosError, AxiosInstance, AxiosRequestConfig } from "axios";
import { Action, Store } from "redux";
import { ThunkDispatch } from "redux-thunk";
import actions from "../actions";
import { Auth } from "./Auth";
import {
  ConflictError,
  ForbiddenError,
  IStoreState,
  NotFoundError,
  UnauthorizedError,
  UnprocessableEntity,
} from "./types";

// createAxiosInterceptors will configure a set of interceptors to a provided axios instance,
// relying also on an external redux store for action dispatching
export function createAxiosInterceptors(axios: AxiosInstance, store: Store<IStoreState>) {
  axios.interceptors.request.use((config: AxiosRequestConfig) => {
    const authToken = Auth.getAuthToken();
    if (authToken) {
      config.headers.Authorization = `Bearer ${authToken}`;
    }
    return config;
  });
  axios.interceptors.response.use(
    response => response,
    e => {
      const dispatch = store.dispatch as ThunkDispatch<IStoreState, null, Action>;
      const err: AxiosError = e;
      if (
        err.config.xsrfCookieName === "XSRF-TOKEN" &&
        err.code === undefined &&
        err.message === "Network Error" &&
        !err.response &&
        Auth.usingOIDCToken()
      ) {
        // The OIDC token is no longer valid, logout
        dispatch(actions.auth.logout());
      }
      let message = err.message;
      if (err.response && err.response.data.message) {
        message = err.response.data.message;
      }
      switch (err.response && err.response.status) {
        case 401:
          // Global action dispatch to log the user out
          dispatch(actions.auth.authenticationError(message));
          dispatch(actions.auth.logout());
          return Promise.reject(new UnauthorizedError(message));
        case 403:
          return Promise.reject(new ForbiddenError(message));
        case 404:
          return Promise.reject(new NotFoundError(message));
        case 409:
          return Promise.reject(new ConflictError(message));
        case 422:
          return Promise.reject(new UnprocessableEntity(message));
        default:
          return Promise.reject(new Error(message));
      }
    },
  );
}
