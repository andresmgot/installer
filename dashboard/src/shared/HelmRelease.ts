import { inflate } from "pako";
import { clearInterval, setInterval } from "timers";

import { App } from "./App";
import { AppRepository } from "./AppRepository";
import { axios } from "./Auth";
import { hapi } from "./hapi/release";
import { IApp, IAppConfigMap, IChart, IChartVersion, IHelmRelease } from "./types";
import * as url from "./url";

export class HelmRelease {
  public static async create(
    helmCRDReleaseName: string,
    tillerReleaseName: string,
    namespace: string,
    chartVersion: IChartVersion,
    values?: string,
  ) {
    const chartAttrs = chartVersion.relationships.chart.data;
    const repo = await AppRepository.get(chartAttrs.repo.name);
    const auth = repo.spec.auth;
    const endpoint = HelmRelease.getResourceLink(namespace);
    const { data } = await axios.post(endpoint, {
      apiVersion: "helm.bitnami.com/v1",
      kind: "HelmRelease",
      metadata: {
        annotations: {
          "apprepositories.kubeapps.com/repo-name": chartAttrs.repo.name,
        },
        name: helmCRDReleaseName,
      },
      spec: {
        auth,
        chartName: chartAttrs.name,
        releaseName: tillerReleaseName,
        repoUrl: chartAttrs.repo.url,
        values,
        version: chartVersion.attributes.version,
      },
    });
    return data;
  }

  public static async upgrade(
    helmCRDReleaseName: string,
    tillerReleaseName: string,
    namespace: string,
    chartVersion: IChartVersion,
    values?: string,
  ) {
    const chartAttrs = chartVersion.relationships.chart.data;
    const repo = await AppRepository.get(chartAttrs.repo.name);
    const auth = repo.spec.auth;
    const endpoint = HelmRelease.getSelfLink(helmCRDReleaseName, namespace);
    const { data } = await axios.patch(
      endpoint,
      {
        apiVersion: "helm.bitnami.com/v1",
        kind: "HelmRelease",
        metadata: {
          name: helmCRDReleaseName,
        },
        spec: {
          auth,
          chartName: chartAttrs.name,
          releaseName: tillerReleaseName,
          repoUrl: chartAttrs.repo.url,
          values,
          version: chartVersion.attributes.version,
        },
      },
      {
        headers: { "Content-Type": "application/merge-patch+json" },
      },
    );
    return data;
  }

  public static async delete(helmCRDReleaseName: string, namespace: string) {
    // strip namespace from release name
    const { data } = await axios.delete(this.getSelfLink(helmCRDReleaseName, namespace));
    return data;
  }

  public static async getAllHelmReleases(namespace?: string) {
    const { data: { items: helmReleaseList } } = await axios.get<{ items: IHelmRelease[] }>(
      this.getResourceLink(namespace),
    );
    return helmReleaseList;
  }

  public static async getHelmRelease(tillerReleaseName: string, namespace: string) {
    const helmReleaseList = await this.getAllHelmReleases(namespace);
    let helmRelease = "";
    helmReleaseList.forEach(r => {
      if (r.spec.releaseName === tillerReleaseName) {
        helmRelease = r.metadata.name;
      }
    });
    return helmRelease;
  }

  public static async getAllWithDetails(namespace?: string) {
    const helmReleaseList = await this.getAllHelmReleases(namespace);
    // Convert list of HelmReleases to release name -> HelmRelease pair
    const helmReleaseMap = helmReleaseList.reduce((acc, hr) => {
      const tillerReleaseName =
        !hr.spec.releaseName || hr.spec.releaseName === ""
          ? `${hr.metadata.name}-${hr.metadata.namespace}`
          : hr.spec.releaseName;
      acc[tillerReleaseName] = hr;
      return acc;
    }, new Map<string, IHelmRelease>());

    // Get the HelmReleaseConfigMaps for all HelmReleases
    const { data: { items: allConfigMaps } } = await axios.get<{
      items: IAppConfigMap[];
    }>(App.getConfigMapsLink());

    // Convert list of HelmReleaseConfigMaps to release name -> latest
    // HelmReleaseConfigMap pair
    const cms = allConfigMaps.reduce((acc, cm) => {
      const releaseName = cm.metadata.labels.NAME;
      // If we've already found a version for this release, only
      // replace it if the version is greater
      if (releaseName in acc) {
        acc[releaseName] = this.getNewest(acc[releaseName], cm);
      } else {
        acc[releaseName] = cm;
      }
      return acc;
    }, new Map<string, IAppConfigMap>());

    // Go through all HelmReleaseConfigMaps and parse as IApp objects
    const apps = Object.keys(cms).map(key => this.parseRelease(cms[key], helmReleaseMap[key]));

    // Fetch charts for each app
    return Promise.all<IApp>(apps.map(async app => this.getChart(app)));
  }

  public static async getDetails(
    helmCRDReleaseName: string,
    tillerReleaseName: string,
    namespace: string,
  ) {
    let hr;
    if (helmCRDReleaseName !== "") {
      const i = await axios.get<IHelmRelease>(this.getSelfLink(helmCRDReleaseName, namespace));
      hr = i.data;
    }
    const items = await this.getDetailsWithRetry(tillerReleaseName);
    // Helm/Tiller will store details in a ConfigMap for each revision,
    // so we need to filter these out to pick the latest version
    const helmConfigMap: IAppConfigMap = items.reduce((ret, cm) => {
      return this.getNewest(ret, cm);
    }, items[0]);

    const app = this.parseRelease(helmConfigMap, hr);
    return await this.getChart(app);
  }

  private static getDetailsWithRetry(tillerReleaseName: string) {
    const getConfigMaps = () => {
      return axios.get<{ items: IAppConfigMap[] }>(App.getConfigMapsLink([tillerReleaseName]));
    };
    return new Promise<IAppConfigMap[]>(async (resolve, reject) => {
      let req = await getConfigMaps();
      if (req.data.items.length > 0) {
        resolve(req.data.items);
        return;
      }
      let retries = 3;
      const t = setInterval(async () => {
        if (retries <= 0) {
          clearInterval(t);
          reject();
        } else {
          req = await getConfigMaps();
          if (req.data.items.length > 0) {
            clearInterval(t);
            resolve(req.data.items);
          }
          retries = retries - 1;
        }
      }, 1000);
    });
  }

  private static getSelfLink(name: string, namespace: string) {
    return `/api/kube/apis/helm.bitnami.com/v1/namespaces/${namespace}/helmreleases/${name}`;
  }

  private static getResourceLink(namespace?: string) {
    if (namespace) {
      return `/api/kube/apis/helm.bitnami.com/v1/namespaces/${namespace}/helmreleases`;
    } else {
      return `/api/kube/apis/helm.bitnami.com/v1/helmreleases`;
    }
  }

  // Takes two IAppConfigMaps and returns the highest version
  private static getNewest(cm1: IAppConfigMap, cm2: IAppConfigMap) {
    const cm1Version = parseInt(cm1.metadata.labels.VERSION, 10);
    const cm2Version = parseInt(cm2.metadata.labels.VERSION, 10);
    return cm1Version > cm2Version ? cm1 : cm2;
  }

  // decode base64, ungzip (inflate) and parse as a protobuf message
  private static parseRelease(cm: IAppConfigMap, hr?: IHelmRelease): IApp {
    const protoBytes = inflate(atob(cm.data.release));
    const rel = hapi.release.Release.decode(protoBytes);
    const app: IApp = { data: rel, type: "helm", hr };
    if (hr && hr.metadata) {
      const repoName = hr.metadata.annotations["apprepositories.kubeapps.com/repo-name"];
      if (repoName) {
        app.repo = {
          name: repoName,
          url: hr.spec.repoUrl,
        };
      }
    }
    return app;
  }

  private static async getChart(app: IApp) {
    if (app.repo && app.data.chart && app.data.chart.metadata) {
      const res = await axios.get<{ data: IChart }>(
        url.api.charts.get(`${app.repo.name}/${app.data.chart.metadata.name}`),
      );
      app.chart = res.data.data;
    }
    return app;
  }
}
