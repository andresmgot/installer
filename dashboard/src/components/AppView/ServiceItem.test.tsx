import { shallow } from "enzyme";
import * as React from "react";

import { IResource } from "shared/types";
import ServiceItem from "./ServiceItem";

it("renders a simple view without IP", () => {
  const service = {
    metadata: {
      name: "foo",
    },
    spec: {
      type: "ClusterIP",
    },
  } as IResource;
  const wrapper = shallow(<ServiceItem service={service} extended={false} />);
  expect(wrapper).toMatchSnapshot();
  expect(wrapper.find(".ServiceItem__url").text()).toBe("None");
});

it("renders a simple view with IP", () => {
  const service = {
    metadata: {
      name: "foo",
    },
    spec: {
      ports: [{ port: 80 }],
      type: "LoadBalancer",
    },
    status: {
      loadBalancer: {
        ingress: [{ ip: "1.2.3.4" }],
      },
    },
  } as IResource;
  const wrapper = shallow(<ServiceItem service={service} extended={false} />);
  expect(wrapper.find(".ServiceItem__url").text()).toBe("http://1.2.3.4");
});

it("renders an extended view with IP", () => {
  const service = {
    metadata: { name: "foo" },
    spec: { ports: [{ port: 80, protocol: "http" }], type: "LoadBalancer" },
    status: { loadBalancer: { ingress: [{ ip: "1.2.3.4" }] } },
  } as IResource;
  const wrapper = shallow(<ServiceItem service={service} extended={true} />);
  expect(wrapper).toMatchSnapshot();
});
