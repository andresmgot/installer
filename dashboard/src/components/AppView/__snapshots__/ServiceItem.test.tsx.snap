// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a simple view without IP 1`] = `
<tr>
  <td>
    foo
  </td>
  <td>
    <a
      href="None"
      key="None"
      target="_blank"
    >
      <span
        className="ServiceItem__url padding-tiny padding-h-normal type-small margin-r-small"
      >
        None
      </span>
    </a>
  </td>
</tr>
`;

exports[`renders an extended view with IP 1`] = `
<tr>
  <td>
    foo
  </td>
  <td>
    LoadBalancer
  </td>
  <td />
  <td>
    1.2.3.4
  </td>
  <td>
    80/http
  </td>
</tr>
`;
